common/tqid/task_queue_id.go

```diff
@@ -0,0 +1,339 @@
+		partitionId int
+	}
+
+	// StickyPartition is made by SDK for a single workflow worker to keep workflow tasks of the same execution
+	// in the same worker for caching benefits. Each sticky partition is identified by a unique `stickyName`
+	// generated by SDK. A StickyPartition can only have workflow task type.
+	StickyPartition struct {
+		stickyName string
+		taskQueue  *TaskQueue
+	}
+
+	// PartitionKey uniquely identifies a task queue partition, to be used in maps.
+	// Note that task queue kind (sticky vs normal) and normal name for sticky task queues are not
+	// part of the task queue partition identity.
+	PartitionKey struct {
+		namespaceID string
+		name        string
+		partitionId int
+		taskType    enumspb.TaskQueueType
+	}
+)
+
+var _ Partition = (*NormalPartition)(nil)
+var _ Partition = (*StickyPartition)(nil)
+
+var (
+	ErrNoParent      = errors.New("root task queue partition has no parent")
+	ErrInvalidDegree = errors.New("invalid task queue partition branching degree")
+	ErrNonZeroSticky = errors.New("only sticky partitions can not have non-zero partition ID")
+)
+
+// FromFamilyName takes a user-provided task queue name (aka family name) and returns a TaskQueueFamily. Returns an
+// error if name looks like a mangled name.
+func FromFamilyName(namespaceId string, name string) (*TaskQueueFamily, error) {
+	if strings.HasPrefix(name, nonRootPartitionPrefix) {
+		return nil, serviceerror.NewInvalidArgument("task queue family name cannot have prefix /_sys/ " + name)
+	}
+	return &TaskQueueFamily{
+		namespaceId: namespace.ID(namespaceId),
+		name:        name,
+	}, nil
+}
+
+// UnsafeTaskQueueFamily should be avoided as much as possible. Use FromFamilyName instead as it validates the tq name.
+func UnsafeTaskQueueFamily(namespaceId string, name string) *TaskQueueFamily {
+	return &TaskQueueFamily{namespace.ID(namespaceId), name}
+}
+
+func FromProto(proto *taskqueuepb.TaskQueue, namespaceId string, taskType enumspb.TaskQueueType) (Partition, error) {
+	baseName, partition, err := parseRpcName(proto.GetName())
+	if err != nil {
+		return nil, err
+	}
+
+	kind := proto.GetKind()
+	normalName := proto.GetNormalName()
+	if normalName != "" && kind != enumspb.TASK_QUEUE_KIND_STICKY {
+		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("only sticky queues can have normal name. tq: %s, normal name: %s", baseName, normalName))
+	}
+
+	switch kind {
+	case enumspb.TASK_QUEUE_KIND_STICKY:
+		if partition != 0 {
+			return nil, fmt.Errorf("%w. base name: %s, normal name: %s", ErrNonZeroSticky, baseName, normalName)
+		}
+		tq := &TaskQueue{TaskQueueFamily{namespace.ID(namespaceId), normalName}, taskType}
+		return tq.StickyPartition(baseName), nil
+	default:
+		tq := &TaskQueue{TaskQueueFamily{namespace.ID(namespaceId), baseName}, taskType}
+		return tq.NormalPartition(partition), nil
+	}
+}
+
+func (n *TaskQueueFamily) Name() string {
+	return n.name
+}
+
+func (n *TaskQueueFamily) NamespaceID() namespace.ID {
```

Comment:
Do we have agreement on proper case for `id`? Id is a short of Identification. It is not abbreviation. So I vote for `Id` (capital `I` lower case `d`)

Reply:
Good point. I'm also not sure what is the convention here. It seems both styles are used in different places of the code. I also like `NamespaceId` better. So changing it according to your suggestion.